x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/huggingface/HuggingFaceElserResponseEntity.java

```diff
@@ -0,0 +1,115 @@
+     *        }
+     *     </code>
+     * </pre>
+     *
+     * The response would look like:
+     *
+     * <pre>
+     *   <code>
+     *     [
+     *       {
+     *         "the": 0.7226026,
+     *         "to": 0.29198948,
+     *         "is": 0.059944477,
+     *         ...
+     *       },
+     *       {
+     *           "wish": 0.123456,
+     *           ...
+     *       }
+     *     ]
+     *   </code>
+     * </pre>
+     */
+    public static TextExpansionResults fromResponse(HttpResult response) throws IOException {
+        var parserConfig = XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+
+        try (XContentParser jsonParser = XContentFactory.xContent(XContentType.JSON).createParser(parserConfig, response.body())) {
+            if (jsonParser.currentToken() == null) {
+                jsonParser.nextToken();
+            }
+
+            List<TextExpansionResults> parsedResponse = XContentParserUtils.parseList(
+                jsonParser,
+                HuggingFaceElserResponseEntity::parseExpansionResult
+            );
+
+            if (parsedResponse.isEmpty()) {
+                return new TextExpansionResults(DEFAULT_RESULTS_FIELD, Collections.emptyList(), false);
+            }
+
+            // we only handle a single response right now so just grab the first one
+            return parsedResponse.get(0);
+        }
+    }
+
+    private static TextExpansionResults parseExpansionResult(XContentParser parser) throws IOException {
+        XContentParser.Token token = parser.currentToken();
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, token, parser);
+
+        List<TextExpansionResults.WeightedToken> weightedTokens = new ArrayList<>();
+
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.FIELD_NAME, token, parser);
+            var floatToken = parser.nextToken();
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.VALUE_NUMBER, floatToken, parser);
+
+            switch (parser.numberType()) {
+                case DOUBLE -> weightedTokens.add(
+                    new TextExpansionResults.WeightedToken(parser.currentName(), (float) parser.doubleValue())
+                );
+                case FLOAT -> weightedTokens.add(new TextExpansionResults.WeightedToken(parser.currentName(), parser.floatValue()));
+                default -> {
+                    logger.warn(format("Received an invalid number type in response [%s]", parser.numberType()));
+                    throw new IllegalArgumentException(
+                        format("Failed to parse object: expecting number token of type float or double but found [%s]", parser.numberType())
+                    );
+                }
+            }
+        }
+        // TODO how do we know if it was truncated?
```

Comment:
Add a test with a truncated document and confirm that the resulting parse exception gets handled nicely. I.e. no closing brace after one of the token/weight objects, and the text representation just ends there.

If a document is valid in terms of matching brackets and braces but is missing some tokens then there'll be no way to detect that. We'll have to trust that the code on the HuggingFace side doesn't do that.